### PR TITLE
[Validator] prevent hash collisions caused by reused object hashes

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #36415
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by xabbuh** [comment]
ping @fancyweb

File: src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php [hunk 1]
```diff
@@ -828,4 +852,23 @@ private function validateInGroup($value, $cacheKey, MetadataInterface $metadata,
...
+     * @return string
+     */
+    private function generateCacheKey($object)
+    {
+        $this->cachedObjectsRefs[] = $object;
+
+        return spl_object_hash($object);
+    }
+
+    private function cleanUpObjectRefs()
```
Comment (stof) [inline src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php]: shouldn't calls to this `cleanUpObjectRefs` happen in a `finally` block instead, to be sure we also record the end of the callstack when some called code triggers an exception ?

Comment (xabbuh) [reply src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php]: In fact keeping the object references in the validator instance does not work as expected as new validator instances can be created in one validation run. I decided to move the reference handling to the execution context instead which also saves us from worrying about have to clean up the stored references instead.

**Review by nicolas-grekas — APPROVED** [review]

**Review by fancyweb — CHANGES_REQUESTED** [review]
As discussed on Slack, we would need a test that assert that using the same constraint instance on several paths on the same `RecursiveContextualValidator` instance (by using `atPath()`) works correctly.

**Review by yceruto — APPROVED** [review]

**Review by fancyweb — APPROVED** [review]

**Comment by xabbuh** [comment]
Thank you @fancyweb.

**Pull request merged**